caffe2/serialize/inline_container.h

```diff
@@ -59,195 +63,271 @@ namespace torch { namespace jit {
+constexpr uint64_t kMaxSupportedFileFormatVersion = 0x1L;
 
-  // Writer-specific constants
-  static constexpr uint64_t kFileFormatVersion = 0x1L;
-  static constexpr uint8_t kPadValue = 0xEF;
+// Writer-specific constants
+constexpr uint64_t kFileFormatVersion = 0x1L;
+constexpr char kPadValue = -17; // 0xEF
 
 }  // namespace
 
-class PyTorchStreamReader {
+class PyTorchStreamReader final {
  public:
-  PyTorchStreamReader(std::istream& in_) : in(in_) {
+  PyTorchStreamReader(std::istream* in) : in_(in) {
     // Store file size so we know when we're done reading because the f* APIs
     // don't do a good job of that
-    in.seekg(0L, in.end);
-    file_size = in.tellg();
-    in.seekg(0L);
-    readAndValidateFileHeader();
+    in_->seekg(0L, in_->end);
+    file_size_ = in_->tellg();
+    readAndValidateFileFooter();
     // Do this now since we're reasonably sure this is actually a PyT file from
     // the header.
-    if (file_size % kFieldAlignment != 0) {
-      throw std::runtime_error("File length is not a multiple of the alignment"
-                               " size. Is this a valid PyTorch file?");
+    if (file_size_ % kFieldAlignment != 0) {
+      throw std::runtime_error(
+          "File length is not a multiple of the alignment"
+          " size. Is this a valid PyTorch model file?");
     }
-    readAndValidateFileFooter();
+    readAndValidateFileHeader();
   }
+
   std::tuple<at::DataPtr, size_t> getLastRecord() {
-    return getRecordWithKey(last_record_offset);
+    return getRecordWithKey(last_record_offset_);
   }
+
   std::tuple<at::DataPtr, size_t> getRecordWithKey(uint64_t key) {
-    if (key + kFieldAlignment > file_size) {
+    if (key + kFieldAlignment > file_size_) {
       throw std::runtime_error("Provided key is larger than the size of the file.");
     }
     if (key % kFieldAlignment != 0) {
       throw std::runtime_error("Provided key is not divisible by the alignment size.");
     }
     // Seek to the provided offset
-    cursor = key;
-    in.seekg(cursor);
+    cursor_ = key;
+    in_->seekg(cursor_);
     auto tag = read64BitIntegerLittleEndian();
     if (tag != RecordTags::STORAGE) {
       throw std::runtime_error("Attempted to read a record of non-storage type");
     }
     auto size = read64BitIntegerLittleEndian();
     seekToNextAlignmentBoundary();
-    auto ptr = malloc(size);
+    auto* ptr = malloc(size);
     at::DataPtr retval(ptr, ptr, free, at::kCPU);
 
-    in.read((char*)ptr, size);
-    cursor += size;
+    in_->read(static_cast<char*>(ptr), size);
+    cursor_ += size;
     seekToNextAlignmentBoundary();
     return std::tuple<at::DataPtr, size_t>(std::move(retval), size);
   }
+
+  // return dataptr, key, size
+  std::tuple<at::DataPtr, int64_t, int64_t> getNextRecord() {
+    int64_t key = cursor_;
+    if (!hasNextRecord()) {
+      throw std::runtime_error("No more record, but hasNextRecord is called.");
+    }
+    auto tag = read64BitIntegerLittleEndian();
```

Comment:
stuff from here down is copy-pasta - have a subfunction?

Reply:
make sense, will do